### PR TITLE
Update Scala to 3.0.0-RC1

### PR DESCRIPTION
**Pull request by julienrf**
I think we should not let the implementations in `dotty-staging` and this one diverge much.

I cherry-picked their commits, and I added a couple of more commits to clean things up.

### CODE REVIEW

File: modules/data/src/main/scala/shapeless3/data/monoidal.scala [hunk 1]
```diff
@@ -18,11 +18,9 @@ package shapeless3.data
 
 import scala.compiletime._
 
-import Monoidal._
-
 trait Monoidal {
-  type to[_]
-  type from[_]
+  type to[_] <: Tuple
+  type from[_ <: Tuple]
```
Comment (joroKr21) [inline modules/data/src/main/scala/shapeless3/data/monoidal.scala:23]: Does it only work with `Tuple` now? 🤔

Comment (julienrf) [reply modules/data/src/main/scala/shapeless3/data/monoidal.scala:23]: I’m not sure, this change was applied in commit 5c4777fbe0f0ee29f1f048acaac300f41df315bf, maybe @milessabin knows more?

**Comment by milessabin** [comment]
Thanks for making a start on this ... I've got some time to move things along next week.

File: modules/data/src/main/scala/shapeless3/data/monoidal.scala (hunk 1 above)
Comment (milessabin) [reply modules/data/src/main/scala/shapeless3/data/monoidal.scala:23]: Tuples are now used as the canonical representation ... there doesn't seem to be any obvious reason to invent another type to play that role.

Comment (joroKr21) [reply modules/data/src/main/scala/shapeless3/data/monoidal.scala:23]: Ok for tuples probably not, but for `Either` the jury is still out I think 🤔

**Comment by milessabin** [comment]
This looks great ... I'm going to merge as-is, then follow up with an additional tweak from @smarter, dependency updates and switch to github actions.

File: modules/data/src/main/scala/shapeless3/data/monoidal.scala (hunk 1 above)
Comment (milessabin) [reply modules/data/src/main/scala/shapeless3/data/monoidal.scala:23]: No, this is fine for `Either` as well ... see [here](https://github.com/milessabin/shapeless/blob/shapeless-3/modules/data/src/main/scala/shapeless3/data/monoidal.scala#L141-L164).

**Pull request merged**